# 

_**DELETE BEFORE PUBLISHING**_

_This is a template also containing the style guide for use cases. The styling uses the use-case css when uploaded to the website, which will not be visible on your local machine._

_Change any text marked with {} and delete any cells marked DELETE_

***

In [5]:
# DELETE BEFORE PUBLISHING
# This is just here so you can preview the styling on your local machine

from IPython.core.display import HTML
HTML("""
<style>
.usecase-title, .usecase-duration, .usecase-section-header {
    padding-left: 15px;
    padding-bottom: 10px;
    padding-top: 10px;
    padding-right: 15px;
    background-color: #0f9295;
    color: #fff;
}

.usecase-title {
    font-size: 1.7em;
    font-weight: bold;
}

.usecase-authors, .usecase-level, .usecase-skill {
    padding-left: 15px;
    padding-bottom: 7px;
    padding-top: 7px;
    background-color: #baeaeb;
    font-size: 1.4em;
    color: #121212;
}

.usecase-level-skill  {
    display: flex;
}

.usecase-level, .usecase-skill {
    width: 50%;
}

.usecase-duration, .usecase-skill {
    text-align: right;
    padding-right: 15px;
    padding-bottom: 8px;
    font-size: 1.4em;
}

.usecase-section-header {
    font-weight: bold;
    font-size: 1.5em;
}

.usecase-subsection-header, .usecase-subsection-blurb {
    font-weight: bold;
    font-size: 1.2em;
    color: #121212;
}

.usecase-subsection-blurb {
    font-size: 1em;
    font-style: italic;
}
</style>
""")

<div class="usecase-title">Air Quality Prediction</div>

<div class="usecase-authors"><b>Authored by: </b> Tharusha Cao</div>

<div class="usecase-duration"><b>Duration:</b> {90} mins</div>

<div class="usecase-level-skill">
    <div class="usecase-level"><b>Level: </b>Intermediate</div>
    <div class="usecase-skill"><b>Pre-requisite Skills: </b>Python, Machine Learning</div>
</div>

<div class="usecase-section-header">Scenario</div>

Air pollution is one of the hot issues that attracted widespread attention from urban and society management. Air quality prediction is to issue an alarm when severe pollution occurs, or pollution concentration exceeds a specific limit, contributing to the measure-taking of relevant departments this is also may be due to development activity, guiding urban socio-economic activities to promote sustainable urban development.

<div class="usecase-section-header">What this use case will teach you</div>

At the end of this use case you will:
- python
- Machine Learning
- Deep Learning
- Data Analysis and visualization skills

<div class="usecase-section-header">{Heading for introduction or background relating to problem}</div>

{Write your introduction here. Keep it concise. We're not after "War and Peace" but enough background information to inform the reader on the rationale for solving this problem or background non-technical information that helps explain the approach. You may also wish to give information on the datasets, particularly how to source those not being imported from the client's open data portal.}



***

_**DELETE BEFORE PUBLISHING**_

## Style guide for use cases

### Headers

For styling within your markdown cells, there are two choices you can use for headers.

1) You can use HTML classes specific to the use case styling:

```<p class="usecase-subsection-header">This is a subsection header.</p>```

<p style="font-weight: bold; font-size: 1.2em;">This is a subsection header.</p>

```<p class="usecase-subsection-blurb">This is a blurb header.</p>```

<p style="font-weight: bold; font-size: 1em; font-style:italic;">This is a blurb header.</p>


2) Or if you like you can use the markdown header styles:

```# for h1```

```## for h2```

```### for h3```

```#### for h4```

```##### for h5```

## Plot colour schemes

General advice:
1. Use the same colour or colour palette throughout your notebook, unless variety is necessary
2. Select a palette based on the type of data being represented
3. Consider accessibility (colourblindness, low vision)

#### 1) If all of your plots only use 1-2 colors use one of the company style colors:

| Light theme | Dark Theme |
|-----|-----|
|<p style="color:#2af598;">#2af598</p>|<p style="color:#08af64;">#08af64</p>|
|<p style="color:#22e4ac;">#22e4ac</p>|<p style="color:#14a38e;">#14a38e</p>|
|<p style="color:#1bd7bb;">#1bd7bb</p>|<p style="color:#0f9295;">#0f9295</p>|
|<p style="color:#14c9cb;">#14c9cb</p>|<p style="color:#056b8a;">#056b8a</p>|
|<p style="color:#0fbed8;">#0fbed8</p>|<p style="color:#121212;">#121212</p>|
|<p style="color:#08b3e5;">#08b3e5</p>||


#### 2) If your plot needs multiple colors, choose an appropriate palette using either of the following tutorials:
- https://seaborn.pydata.org/tutorial/color_palettes.html
- https://matplotlib.org/stable/tutorials/colors/colormaps.html

#### 3) Consider accessibility as well.

For qualitative plotting Seaborn's 'colorblind' palette is recommended. For maps with sequential or diverging it is recommended to use one of the Color Brewer schemes which can be previewed at https://colorbrewer2.org/.

If you want to design your own colour scheme, it should use the same principles as Cynthia Brewer's research (with variation not only in hue but also, saturation or luminance).

### References

Be sure to acknowledge your sources and any attributions using links or a reference list.

If you have quite a few references, you might wish to have a dedicated section for references at the end of your document, linked using footnote style numbers.

You can connect your in-text reference by adding the number with a HTML link: ```<a href="#fn-1">[1]</a>```

and add a matching ID in the reference list using the ```<fn>``` tag: ```<fn id="fn-1">[1] Author (Year) _Title_, Publisher, Publication location.</fn>```

import requests
import pandas as pd

# Function that extracts the data from the JSON 
def get_data(base, data_url, offset = 0):
    
    filters = f'records?limit={100}&offset={offset}&timezone=UTC'
    url = f'{base}{data_url}/{filters}'
    result = requests.get(url)
    # if you need to extract more than 9900 records, enter the api key like below
    #result = requests.get(url + f'&apikey={API_KEY}')
    result_json = result.json()
    max_results = result_json['total_count']
    links = result_json['links']
    records = result_json['records']
    records_df = pd.json_normalize(records)
    
    # fix columns names
    records_df.drop(columns=['links'],inplace=True)
    column_names = records_df.columns.values.tolist()
    
    column_names = [i.split('.')[-1] for i in column_names]
    records_df.columns = column_names
    
    next_url = None
    
    # get next url
    if records_df.shape[0] != max_results:
        for l in links:
            if l['rel']=='next':
                next_url = l['href']
                
    return [records_df, next_url, column_names]

# Code

Note the "Discussion" sections are displayed under relevant code outputs.

### Fetching Data ###

In [1]:
###################################################################
# Libraries used:
###################################################################
import numpy as np
import pandas as pd
import seaborn as sns
import folium
import matplotlib.pyplot as plt
import requests
import math
from geopy.distance import geodesic
from folium.plugins import MarkerCluster
from sklearn.preprocessing import MinMaxScaler


In [2]:
ArgyleSquareAirQuality = "https://data.melbourne.vic.gov.au/api/explore/v2.1/catalog/datasets/argyle-square-air-quality/exports/csv?lang=en&timezone=Australia%2FSydney&use_labels=true&delimiter=%2C"
DevelopmentActivityMonitor ="https://data.melbourne.vic.gov.au/api/explore/v2.1/catalog/datasets/development-activity-monitor/exports/csv?lang=en&timezone=Australia%2FSydney&use_labels=true&delimiter=%2C"
AirQuality_df = pd.read_csv(ArgyleSquareAirQuality)
ActivityMonitor_df = pd.read_csv(DevelopmentActivityMonitor)


In [5]:
pd.set_option('display.max_columns', None)
AirQuality_df.head()

date_measure    dev_id           Sensor Name  \
0  2022-05-04T07:46:34+10:00  ems-ec8a  Air Quality Sensor 2   
1  2021-05-15T16:04:33+10:00  ems-ec8a  Air Quality Sensor 2   
2  2021-05-15T16:10:49+10:00  ems-ce10  Air Quality Sensor 1   
3  2021-05-15T17:49:32+10:00  ems-ec8a  Air Quality Sensor 2   
4  2021-05-15T19:19:32+10:00  ems-ec8a  Air Quality Sensor 2   

                  Lat Long  averageSpl  carbonMonoxide  humidity  iBatt  \
0  -37.802772, 144.9655513        58.0         -6261.0      77.0  169.0   
1  -37.802772, 144.9655513        56.0         -6261.0      51.0   12.0   
2  -37.802772, 144.9655513        70.0         -2887.0      61.0  132.0   
3  -37.802772, 144.9655513        56.0         -6822.0      59.0   98.0   
4  -37.802772, 144.9655513        57.0         -6822.0      63.0  109.0   

   nitrogenDioxide  ozone  particulatesErr  particulatesVsn  peakSpl  pm1  \
0            268.0  137.0              0.0            151.0     64.0  0.0   
1            258.0  119.0              0.0            151.0     62.0  0.0   
2            123.0  -13.0              0.0            151.0     87.0  0.0   
3            306.0  147.0              0.0            151.0     60.0  0.0   
4            306.0  142.0              0.0            151.0     65.0  2.0   

   pm10  pm25  temperature  vBatt  vPanel  
0   0.0   0.0         15.1   3.76   16.33  
1   0.0   0.0         14.9   4.01   18.33  
2   0.0   0.0         13.2   3.90   16.03  
3   1.0   1.0         12.7   3.95    0.00  
4   5.0   3.0         12.1   3.93    0.00

### Pre-Proccessing Data ###
##### This section will consist of pre-processing the data sets such as removing unnecessary columns, Filling missing values , Normalization, Encoding and more to help make the data usefull for machine learning and deep learning 

In [7]:
# Removing Uneccerrary columns which wont be any use for data analysis
#From the Air Quality we will be using carbonMonoxide as the Target Feature 
Droping_Features_AirQuality = ['vPanel','vBatt','pm25','pm10','pm1','peakSpl','iBatt','averageSpl','Sensor Name','dev_id']
AirQuality_df = AirQuality_df.drop(columns=Droping_Features_AirQuality)

In [8]:
AirQuality_df.head()

date_measure                 Lat Long  carbonMonoxide  \
0  2022-05-04T07:46:34+10:00  -37.802772, 144.9655513         -6261.0   
1  2021-05-15T16:04:33+10:00  -37.802772, 144.9655513         -6261.0   
2  2021-05-15T16:10:49+10:00  -37.802772, 144.9655513         -2887.0   
3  2021-05-15T17:49:32+10:00  -37.802772, 144.9655513         -6822.0   
4  2021-05-15T19:19:32+10:00  -37.802772, 144.9655513         -6822.0   

   humidity  nitrogenDioxide  ozone  particulatesErr  particulatesVsn  \
0      77.0            268.0  137.0              0.0            151.0   
1      51.0            258.0  119.0              0.0            151.0   
2      61.0            123.0  -13.0              0.0            151.0   
3      59.0            306.0  147.0              0.0            151.0   
4      63.0            306.0  142.0              0.0            151.0   

   temperature  
0         15.1  
1         14.9  
2         13.2  
3         12.7  
4         12.1

#### In this first part we will be trying to predict future AirQuality in Argyle Square, The features used :


##### 1. date_measure - Important component as we need to recognize the change in the pattern
##### 2. Lat Long - Currently only needed for part 2
##### 3. carbonMonoxide - This is the Target for this part, we need to find the carbonMonoxide value for future Dates
##### 4. humidity - Training with corelation
##### 5. nitrogenDioxide - Training with corelation
##### 6. ozone - Training with corelation
##### 7. temperature - Training with corelation
##### 8. particulatesErr - Will be used in Part 2
##### 9. particulatesVsn - Will be used in Part 2

In [19]:
#Checking for empty data 
missing_values = AirQuality_df.isnull().sum()
print("--------------------------Current Missing Values --------------------------")
print(missing_values)


#Lets see propotion
missing_proportion = (missing_values / len(AirQuality_df))*100
print("--------------------------Current Missing Values Propotion --------------------------")
print(missing_proportion)




--------------------------Current Missing Values --------------------------
date_measure          0
Lat Long              0
carbonMonoxide     9765
humidity           9765
nitrogenDioxide    9765
ozone              9765
particulatesErr    9765
particulatesVsn    9765
temperature        9765
dtype: int64
--------------------------Current Missing Values Propotion --------------------------
date_measure       0.000000
Lat Long           0.000000
carbonMonoxide     6.993583
humidity           6.993583
nitrogenDioxide    6.993583
ozone              6.993583
particulatesErr    6.993583
particulatesVsn    6.993583
temperature        6.993583
dtype: float64


##### We can see that there are around 9765 missing values for the Above and We can see that the proportion is 6.993583% of 139629 data points in Aquality Data set 